// fishy-bot-framework/lib/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FishyClient = void 0;
const client_1 = require("./client");
Object.defineProperty(exports, "FishyClient", { enumerable: true, get: function () { return client_1.FishyClient; } });
// Usefull for quick testing, only gets run when started with `node index.js`
if (require.main === module) {
    (() => __awaiter(void 0, void 0, void 0, function* () {
        require("dotenv").config();
        let { GuildModel } = require("./test/models/Guild");
        let client = new client_1.FishyClient({
            token: process.env.discord_token,
            db_uri: process.env.mongodb_uri,
            guild_model: GuildModel,
            author: "Fish#2455",
            cmd_dir: "./lib/test/commands",
            disable_load_on_construct: true,
        });
        yield client.load();
        yield client.login();
        console.log(client.categories);
    }))();
}
