// fishy-bot-framework/lib/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InteractionType = exports.ApplicationCommandOptionType = exports.InteractionApplicationCommandCallbackDataFlags = exports.InteractionResponseType = void 0;
var InteractionResponseType;
(function (InteractionResponseType) {
    InteractionResponseType[InteractionResponseType["Pong"] = 1] = "Pong";
    InteractionResponseType[InteractionResponseType["ChannelMessageWithSource"] = 4] = "ChannelMessageWithSource";
    InteractionResponseType[InteractionResponseType["DeferredChannelMessageWithSource"] = 5] = "DeferredChannelMessageWithSource";
})(InteractionResponseType = exports.InteractionResponseType || (exports.InteractionResponseType = {}));
var InteractionApplicationCommandCallbackDataFlags;
(function (InteractionApplicationCommandCallbackDataFlags) {
    InteractionApplicationCommandCallbackDataFlags[InteractionApplicationCommandCallbackDataFlags["None"] = 0] = "None";
    InteractionApplicationCommandCallbackDataFlags[InteractionApplicationCommandCallbackDataFlags["Ephemeral"] = 256] = "Ephemeral";
})(InteractionApplicationCommandCallbackDataFlags = exports.InteractionApplicationCommandCallbackDataFlags || (exports.InteractionApplicationCommandCallbackDataFlags = {}));
var ApplicationCommandOptionType;
(function (ApplicationCommandOptionType) {
    ApplicationCommandOptionType[ApplicationCommandOptionType["SUB_COMMAND"] = 1] = "SUB_COMMAND";
    ApplicationCommandOptionType[ApplicationCommandOptionType["SUB_COMMAND_GROUP"] = 2] = "SUB_COMMAND_GROUP";
    ApplicationCommandOptionType[ApplicationCommandOptionType["STRING"] = 3] = "STRING";
    ApplicationCommandOptionType[ApplicationCommandOptionType["INTEGER"] = 4] = "INTEGER";
    ApplicationCommandOptionType[ApplicationCommandOptionType["BOOLEAN"] = 5] = "BOOLEAN";
    ApplicationCommandOptionType[ApplicationCommandOptionType["USER"] = 6] = "USER";
    ApplicationCommandOptionType[ApplicationCommandOptionType["CHANNEL"] = 7] = "CHANNEL";
    ApplicationCommandOptionType[ApplicationCommandOptionType["ROLE"] = 8] = "ROLE";
})(ApplicationCommandOptionType = exports.ApplicationCommandOptionType || (exports.ApplicationCommandOptionType = {}));
var InteractionType;
(function (InteractionType) {
    InteractionType[InteractionType["Ping"] = 1] = "Ping";
    InteractionType[InteractionType["ApplicationCommand"] = 2] = "ApplicationCommand";
})(InteractionType = exports.InteractionType || (exports.InteractionType = {}));
