// fishy-bot-framework/lib/test/commands/info/ping.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.help = exports.config = exports.run = void 0;
const run = (client, interaction) => __awaiter(void 0, void 0, void 0, function* () {
    console.log(yield interaction.getDbGuild());
    interaction.send("asddfdafadf");
});
exports.run = run;
exports.config = {
    name: "ping",
    bot_needed: false,
    interaction_options: {
        name: "ping",
        description: "Ping the bot"
    },
};
exports.help = {
    description: "bbblah",
    usage: "blah",
};
